compiler/rustc_trait_selection/src/infer.rs

```diff
@@ -79,6 +98,39 @@ impl<'cx, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'cx, 'tcx> {
         );
         InferOk { value, obligations }
     }
+
+    fn type_implements_trait(
+        &self,
+        trait_def_id: DefId,
+        ty: Ty<'tcx>,
+        params: SubstsRef<'tcx>,
+        param_env: ty::ParamEnv<'tcx>,
+    ) -> traits::EvaluationResult {
+        debug!(
+            "type_implements_trait: trait_def_id={:?}, type={:?}, params={:?}, param_env={:?}",
+            trait_def_id, ty, params, param_env
+        );
+
+        let trait_ref =
+            ty::TraitRef { def_id: trait_def_id, substs: self.tcx.mk_substs_trait(ty, params) };
+
+        // FIXME(#86868): If there are inference variables anywhere, just assume
+        // we don't know the answer. This works around the ICEs that would result
+        // from using those inference variables within the `infer_ctxt` we create below.
+        // Really this should be a method on `InferCtxt` instead, which would fix this
+        // problem.
+        if (trait_ref, param_env).needs_infer() {
+            return traits::EvaluationResult::EvaluatedToUnknown;
+        }
```

Comment:
```suggestion
```

We don't need this anymore! The problem was that we were creating a fresh inference context, and hence those type variables and things were out of scope in that fresh infrence context. Since we're not doing that, it's the callers problem to pass up types that are appropriate for this inference context.